src/control/layer/LayerController.h

```diff
@@ -106,6 +107,10 @@ class LayerController: public DocumentListener {
      */
     void ensureLayerExists(PageRef page);
 
+    // Allows us to rebuild the layer menu from the undo action class upon
+    // redoing a merge.
+    friend class MergeLayerDownUndoAction;
```

Comment:
I think it might make more sense to add a public `rebuildLayerMenu` or `fireLayerUpdate` method to `LayerControler` than adding `MergeLayerDownUndoAction` as a `friend` class... I'm not sure though...

Reply:
Definitely do not use a friend class here. While some people think this is okay, it makes understanding the code really hard and is not worth the "added protection" of not having this method be public.

Reply:
Just as a follow-up for clarification:
From the top of my mind, there are only a few very limited cases where you would be interested in using a friend class. And those are really specific cases like:

1. An iterator for multiple custom classes that must access private state within these classes
2. A really big class you want to split up into smaller units for maintenance but treat them as one big class from the outside
3. A unit test requires a check of an internal state and you really have no good way of giving it access without going against common sense (and this is really already a code smell since your unit tests should only ever cover the public interfaces of your units - that is their whole point)

So whenever you happen to stumble upon a friend class/function, you most likely look either at a quick-and-dirty hack around encapsulation or a system that has grown over time, and refactoring would equal rewriting the full codebase.

And while risking an angry mob outside my window: No friend classes do not mirror the concept of `protected` from Java... Whoever got that idea out into the world...